### PR TITLE
PSIBackgroundSubtraction - Add properties StartX and EndX

### LINKED ISSUE #30901: PSIBackgroundSubtraction algorithm - Add StartX, EndX and Function Properties
### Description
We should have the option to specify StartX, EndX and Function for the `PSIBackgroundSubtraction` algorithm.

From Anthony:
```
could you have a look at adding the start and end values as optional arguments to the 
PSIBackgroundCorrection alg, and add an optional input for a function (which we just add to the default) 
and then return the fitted function (I think we may have to do it in full)
```

I have split this into two tasks:
- [ ] Add StartX and EndX properties to `PSIBackgroundSubtraction` algorithm
- [ ] Add an optional Function property that we add onto the default function, then return the full function as an output property

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR adds a `StartX` and `EndX` property to the PSIBackgroundSubtraction algorithm.  The background required to be subtracted will be calculated based on the X values between the `StartX` and `EndX`. The calculated background will then be subtracted across all X's.

**To test:**
Use this script for testing
```py
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = CreateSampleWorkspace()
ws.setYUnit("Counts")

n_bins = ws.blocksize()
n_hist = ws.getNumberHistograms()

for i in range(n_hist):
    ws.mutableRun().addProperty("First good spectra " + str(i),  int(n_bins / 2), True)
    ws.mutableRun().addProperty("Last good spectra " + str(i),  n_bins, True)

PSIBackgroundSubtraction(InputWorkspace=ws, StartX=5000, EndX=6000)
```

1. Try StartX=5000 and EndX=6000. The resulting workspace should all be 0.3's (apart from the peak 10.3 at bin index 50). This is because the background was calculated based on there only being 0.3's between X=5000 and 6000. (meaning the background subtracted was all zero).

2. Try StartX=5000 and EndX=15000. This range encompasses the peak at bin index 50. Therefore a different background is calculated and then subtracted. The resulting workspace should all be approx -0.0058 .

3. Also if you don't provide StartX and EndX we should expect the same result as step 2. (i.e. -0.0058)

Fixes #30901

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Muon/src/PSIBackgroundSubtraction.cpp
Comment (StephenSmith25) [inline Framework/Muon/src/PSIBackgroundSubtraction.cpp:31]: Just wondering why these shared_ptrs are now passed by value? Not 100% clear on shared_ptr semantics still, but my understanding is:

- Pass by value `(std::shared_ptr<A> a)` if you want to share ownership i.e increment the use_count
- Pass by reference `(std::shared_ptr<A>& a)` if you're not trying to share ownership (but may want to reset the pointer)
- Pass by const reference `(const std::shared_ptr<A>& a)`  if you don't want the function to be able to reset the shared ptr (which is what we had before) and don't extend the lifetime i.e. don't increment the use count

With the change, we are now passing by value `shared_ptr<const MatrixWorkspace>` meaning we are sharing ownership, i,e. incrementing the use count, but not letting the function modify the workspace. Is this the desired outcome? 

 I probably wrote the original function, so likely to blame here, but could getRangeFromWorkspace just take a bare reference e.g.

`getRangeFromWorkspace(const MatrixWorkspace& inputWorkspace, const size_t index)`

That way we don't have to care about the ownership debates that come with shared_ptrs.

**Review by StephenSmith25 — COMMENTED** [review]
Looks good and great addition of another test. Just one query and I'll test it once I get my build working again..

File: Framework/Muon/src/PSIBackgroundSubtraction.cpp
Comment (robertapplin) [reply Framework/Muon/src/PSIBackgroundSubtraction.cpp]: Thanks for taking a look, shared_ptr semantics are confusing!

**Comment by StephenSmith25** [comment]
I'm going to unassigned myself as I don't have a working build of mantid nor a machine to test this on, so hopefully someone see can pick this up and test it.

**Comment by Matt-Cumber** [comment]
@robertapplin Does this need a full code review or just testing?

**Comment by robertapplin** [comment]
> @robertapplin Does this need a full code review or just testing?

I think @StephenSmith25 has already done a code review but feel free to talk a look anyways

**Review by Matt-Cumber — CHANGES_REQUESTED** [review]
I have just been testing this PR and I tried startx greater than endx. It does print the correct warning however there is also still a red error message that is shown as well. 
```
Invalid value for EndX: EndX must be greater than StartX.
Invalid value for StartX: StartX must be less than EndX.

RuntimeError: PSIBackgroundSubtraction-v1: Some invalid Properties found: [ EndX StartX ]
  File "<string>", line 16, in <module>
  File "c:\...\mantid\framework\pythoninterface\mantid\simpleapi.py", line 1049, in __call__
    raise RuntimeError(msg) from e
```

**Comment by robertapplin** [comment]
> I have just been testing this PR and I tried startx greater than endx. It does print the correct warning however there is also still a red error message that is shown as well.
> 
> ```
> Invalid value for EndX: EndX must be greater than StartX.
> Invalid value for StartX: StartX must be less than EndX.
> 
> RuntimeError: PSIBackgroundSubtraction-v1: Some invalid Properties found: [ EndX StartX ]
>   File "<string>", line 16, in <module>
>   File "c:\...\mantid\framework\pythoninterface\mantid\simpleapi.py", line 1049, in __call__
>     raise RuntimeError(msg) from e
> ```

Thanks for looking. The same message is shown for all algorithms so I don't think it is related to this PR

**Review by Matt-Cumber — APPROVED** [review]
Everything works as described. Nice new test. Great work.

**Comment by DavidFair** [comment]
@robertapplin I notice this has multiple reviewers requested, do you want me to wait for a co-reviewer before a gatekeeper review + merge? Alternatively, we can look at getting this merged today

**Comment by robertapplin** [comment]
> @robertapplin I notice this has multiple reviewers requested, do you want me to wait for a co-reviewer before a gatekeeper review + merge? Alternatively, we can look at getting this merged today

It should be ok with just one review, thanks for checking!

**Pull request merged**
